torch/distributed/pipelining/schedules.py

```diff
@@ -743,6 +753,76 @@ def _step_microbatches(
         self._update_losses(self._stage, losses)
 
 
+def _add_unshard_reshard(
+    compute_actions: List[Optional[_Action]],
+    max_active_stages: int = 3,
+) -> List[Optional[_Action]]:
+    """Given a basic schedule involving only compute actions (F,B,W), add UNSHARD/RESHARD actions for FSDP.
+
+    UNSHARD refers to fetching the full contents of an FSDP-sharded layer, requiring an all-gather operation.
+    RESHARD does the opposite, releasing memory (but doing no commmunication)
+
+    We abandon the "timestep lock"  during lowering
+
+    max_active_stages controls how many prefetches we allow. It should be measured in mb and tuneable but in practice
+    3 stages is probably the thing we want?
+    (to account for having one f and one b active, and something else prefetching?)
```

Comment:
what did you mean by the "something else prefetching" comment?

Reply:
i'm not sure. i can't even remember why i thought we need 3 instead of 2 now. :/